doc/xgboost_on_sqlflow.md

```diff
@@ -0,0 +1,139 @@
+COLUMN
+  NUMERIC(c1) as fc1,
+  BUCKETIZED(fc1, [0, 10, 100]) as fc2,
+  CROSS([fc1, fc2, c3]) as fc3,
+  c4 as fc4
+LABEL class
+INTO sqlflow_models.xgboost_model_table;
+
+// sample clause of predict
+select 
+    c1, c2, c3, c4
+from kaggle_credit_fraud_development_data
+PREDICT kaggle_credit_fraud_development_data.class
+USING sqlflow_models.xgboost_model_table
+COLUMN
+  NUMERIC(c1) as fc1,
+  BUCKETIZED(fc1, [0, 10, 100]) as fc2,
+  CROSS([fc1, fc2, c3]) as fc3,
+  c4 as fc4;
+```
+
+### Implementation
+
+As `codegen.go` generating _TensorFlow_ code from sqlflow AST,
+we will add `codegen_xgboost.go` which translate sqlflow AST into a python launcher program of _xgboost_. 
+
+In _ant-xgboost_, there exists an incubating module named [_xgblauncher_](https://github.com/alipay/ant-xgboost/tree/ant_master/xgboost-launcher), 
+an extendable, cloud-native xgboost based machine learning pipeline, 
+which provides a python API for building custom `DataSource` and `ModelSource`.
+
+Below is a demonstration of DataSource/ModelSource API.
+ 
+```python
+class DataSource:
+    """
+    DataSource API
+    A handler of data reading/writing, which is compatible with both single-machine and distributed runtime.
+    """
+    def __init__(self, 
+                 rank: int, 
+                 num_worker: int,
+                 column_conf: configs.ColumnFields,
+                 source_conf):
+        pass
+        
+    @abstractmethod
+    def read(self) -> Iterator[XGBoostRecord]:
+        pass
+
+    @abstractmethod
+    def write(self, result_iter: Iterator[XGBoostResult]):
+        pass
+
+    
+class ModelSource:
+    """
+    ModelSource API
+    A handler by which XGBLauncher save/load model(booster) and related information.
+    """
+    def __init__(self, source_conf):
+        pass
+
+    @abstractmethod
+    def read_buffer(self, model_path: str) -> bytes:
+        pass
+
+    @abstractmethod
+    def write_buffer(self, buf: bytes, model_path: str):
+        pass
+
+    @abstractmethod
+    def read_lines(self, model_path: str) -> List[str]:
+        pass
+
+    @abstractmethod
+    def write_lines(self, lines: List[str], model_path: str):
+        pass
+``` 
+
+
+With _xgblauncher_, we can launch _xgboost_ from sqlflow AST via a lightweight python `code template` and a corrsponding `filler`.
+The `code template` roughly includes  components as follows: 
+
+* `TFDataSource` that is responsible for fetching and pre-processing data via tf.feature_columns API.
+   Data will be fetched in mini-batch style by executing TF compute graph with mini-batch data feed by `sqlflow_submitter.db.db_generator`.
+
+* `DBDataSource` that is responsible for writing prediction results into specific data base.
+   The writing action can be implemented via `sqlflow_submitter.db.insert_values`.
+
+* `LocalModelSource` that is responsible for reading/writing _xgboost_ models on local file system.
+
+* Configuration conversions and entry point of _xgblauncher_.
+
+
```

Comment:
Maybe add a "Distributed training" section.

Reply:
@sperlingxx , I found [xgboost operator](https://github.com/kubeflow/xgboost-operator), do you have any experience on it? Or maybe you have some best practices about distributed xgboost?

Reply:
@Yancey1989 xgboost operator is good for distributed xgboost training and batch prediction. xgboost operator would help sqlflow-xgboost distributed xgboost training on k8s.